### ISSUE TITLE
Demo App: Programmatic Marketing / Super-fast auctions 

### ISSUE THREAD

**Issue by stonecoldpat** [counterfactual/monorepo#1162]
To really demonstrate the efficiency / speed / trust-freeness of state channels, I'd like to see someone build a N party auction. 

* Set up - All parties deposit large amount of coins into the channel. 
* Off-chain - All parties co-operatively perform a commit + reveal auction protocol. i.e. auctioneer offers an item to bid on. All parties commit a bid, and then all parties reveal a bid. 

It'd be great if this auction protocol can support asynchronous deposits / withdrawals, where parties can add more coins in (and take coins out) of the channel while it is in operation and the auction is running. The Sprites payment channel could facilitate that part. 

Why is this a cool idea?
https://www.smartinsights.com/internet-advertising/internet-advertising-targeting/what-is-programmatic-marketing/

While most people in our space hate the idea of adverts, it is really interesting that there are real-time and super-fast auctions happening when someone visits a website. Cryptocurrencies provide a super advantage compared to auctions run today:
- Permissionless: Anyone can join / deposit coins, no need to use visa/any payment service provider. 
- Optimistic fair exchange: Assuming auctioneer honours the winner of an auction, it will encourage continuous auctions. 

The biggest downside to such an auction is that 1 party could withhold/stop its progress. But they can just be kicked out of the channel, and it'd be nice to see if someone can think of a fair way to do that. i.e. N-1 parties sign to remove that party. and that authorises a withdrawal on their behalf / blacklists them from the contract. I've not seen many thoughts around this problem, so this is a nice way to explore it.

**Comment by emansipater** [comment]
Yeah, I could see this being a fun example. There are lots of little touches you can add, like making the winner anonymous, doing a Vickrey auction, or stuff like that. I have some pretty good ideas of how you could do it so that no one can really hold up the progress of the auction itself. Auctions have a nice "liveness" requirement that only the people who are actually online at the moment have a right to even participate, which makes it easy to shorten timeouts, etc. and just make it so that people who stop responding lose the auction without losing their money, provided they eventually show up to claim it along with all of their unrevealed secrets or whatever.
